File: tools/scripts-executor/executor.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// ScriptConfig definierar ett script och dess dependencies
type ScriptConfig struct {
	Name         string   `json:"name"`
	Type         string   `json:"type"` // "python" eller "powershell"
	ScriptPath   string   `json:"script_path"`
	PythonDeps   []string `json:"python_deps,omitempty"`
	PSModules    []string `json:"ps_modules,omitempty"`
	WorkingDir   string   `json:"working_dir,omitempty"`
	Args         []string `json:"args,omitempty"`
}

// ExecutorConfig konfiguration för executor
type ExecutorConfig struct {
	PythonPath    string        `json:"python_path,omitempty"`
	PythonEnv     string        `json:"python_env,omitempty"` // Path to venv
	PowerShellPath string        `json:"powershell_path,omitempty"`
	Scripts       []ScriptConfig `json:"scripts"`
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <config.json> [script_name] [args...]\n", os.Args[0])
		os.Exit(1)
	}

	configPath := os.Args[1]
	config, err := loadConfig(configPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error loading config: %v\n", err)
		os.Exit(1)
	}

	// Om script_name anges, kör bara det scriptet
	if len(os.Args) >= 3 {
		scriptName := os.Args[2]
		scriptArgs := os.Args[3:]
		
		err := runScript(config, scriptName, scriptArgs)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error running script: %v\n", err)
			os.Exit(1)
		}
		return
	}

	// Annars kör alla scripts i ordning
	for _, script := range config.Scripts {
		fmt.Printf("Running script: %s\n", script.Name)
		err := executeScript(config, script, nil)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error running %s: %v\n", script.Name, err)
			os.Exit(1)
		}
	}
}

func loadConfig(path string) (*ExecutorConfig, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var config ExecutorConfig
	err = json.Unmarshal(data, &config)
	if err != nil {
		return nil, err
	}

	// Set defaults
	if config.PythonPath == "" {
		config.PythonPath = "python3"
	}
	if config.PowerShellPath == "" {
		config.PowerShellPath = "pwsh"
	}

	return &config, nil
}

func runScript(config *ExecutorConfig, scriptName string, args []string) error {
	for _, script := range config.Scripts {
		if script.Name == scriptName {
			return executeScript(config, script, args)
		}
	}
	return fmt.Errorf("script '%s' not found", scriptName)
}

func executeScript(config *ExecutorConfig, script ScriptConfig, args []string) error {
	var cmd *exec.Cmd
	workingDir := script.WorkingDir
	if workingDir == "" {
		workingDir = filepath.Dir(script.ScriptPath)
	}

	switch strings.ToLower(script.Type) {
	case "python":
		cmd = buildPythonCommand(config, script, args, workingDir)
	case "powershell":
		cmd = buildPowerShellCommand(config, script, args, workingDir)
	default:
		return fmt.Errorf("unknown script type: %s", script.Type)
	}

	cmd.Dir = workingDir
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Stdin = os.Stdin

	return cmd.Run()
}

func buildPythonCommand(config *ExecutorConfig, script ScriptConfig, args []string, workingDir string) *exec.Cmd {
	var pythonExe string
	
	// Använd virtualenv om specificerad
	if config.PythonEnv != "" {
		if _, err := os.Stat(config.PythonEnv); err == nil {
			// Windows: Scripts\python.exe, Linux/Mac: bin/python
			if _, err := os.Stat(filepath.Join(config.PythonEnv, "Scripts", "python.exe")); err == nil {
				pythonExe = filepath.Join(config.PythonEnv, "Scripts", "python.exe")
			} else if _, err := os.Stat(filepath.Join(config.PythonEnv, "bin", "python")); err == nil {
				pythonExe = filepath.Join(config.PythonEnv, "bin", "python")
			}
		}
	}
	
	if pythonExe == "" {
		pythonExe = config.PythonPath
	}

	cmdArgs := []string{script.ScriptPath}
	if len(script.Args) > 0 {
		cmdArgs = append(cmdArgs, script.Args...)
	}
	if len(args) > 0 {
		cmdArgs = append(cmdArgs, args...)
	}

	return exec.Command(pythonExe, cmdArgs...)
}

func buildPowerShellCommand(config *ExecutorConfig, script ScriptConfig, args []string, workingDir string) *exec.Cmd {
	cmdArgs := []string{"-File", script.ScriptPath}
	if len(script.Args) > 0 {
		cmdArgs = append(cmdArgs, script.Args...)
	}
	if len(args) > 0 {
		cmdArgs = append(cmdArgs, args...)
	}

	return exec.Command(config.PowerShellPath, cmdArgs...)
}

File: scripts-executor/executor.rs
use serde::{Deserialize, Serialize};
use std::process::{Command, Stdio};
use std::path::PathBuf;
use std::env;

#[derive(Debug, Deserialize)]
struct ScriptConfig {
    name: String,
    #[serde(rename = "type")]
    script_type: String,
    script_path: String,
    #[serde(default)]
    python_deps: Vec<String>,
    #[serde(default)]
    ps_modules: Vec<String>,
    #[serde(default)]
    working_dir: Option<String>,
    #[serde(default)]
    args: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct ExecutorConfig {
    #[serde(default = "default_python_path")]
    python_path: String,
    python_env: Option<String>,
    #[serde(default = "default_powershell_path")]
    powershell_path: String,
    scripts: Vec<ScriptConfig>,
}

fn default_python_path() -> String {
    "python3".to_string()
}

fn default_powershell_path() -> String {
    "pwsh".to_string()
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    
    if args.len() < 2 {
        eprintln!("Usage: {} <config.json> [script_name] [args...]", args[0]);
        std::process::exit(1);
    }

    let config_path = &args[1];
    let config_str = std::fs::read_to_string(config_path)?;
    let config: ExecutorConfig = serde_json::from_str(&config_str)?;

    // Om script_name anges, kör bara det scriptet
    if args.len() >= 3 {
        let script_name = &args[2];
        let script_args: Vec<String> = args[3..].to_vec();
        
        run_script(&config, script_name, &script_args)?;
        return Ok(());
    }

    // Annars kör alla scripts i ordning
    for script in &config.scripts {
        println!("Running script: {}", script.name);
        execute_script(&config, script, &[])?;
    }

    Ok(())
}

fn run_script(config: &ExecutorConfig, script_name: &str, args: &[String]) -> Result<(), Box<dyn std::error::Error>> {
    for script in &config.scripts {
        if script.name == script_name {
            return execute_script(config, script, args);
        }
    }
    Err(format!("Script '{}' not found", script_name).into())
}

fn execute_script(config: &ExecutorConfig, script: &ScriptConfig, args: &[String]) -> Result<(), Box<dyn std::error::Error>> {
    let working_dir = script.working_dir.as_ref()
        .map(|d| PathBuf::from(d))
        .unwrap_or_else(|| {
            PathBuf::from(&script.script_path)
                .parent()
                .unwrap_or(PathBuf::from(".").as_path())
                .to_path_buf()
        });

    let mut cmd = match script.script_type.to_lowercase().as_str() {
        "python" => build_python_command(config, script, args, &working_dir)?,
        "powershell" => build_powershell_command(config, script, args, &working_dir)?,
        _ => return Err(format!("Unknown script type: {}", script.script_type).into()),
    };

    cmd.current_dir(&working_dir);
    cmd.stdin(Stdio::inherit());
    cmd.stdout(Stdio::inherit());
    cmd.stderr(Stdio::inherit());

    let status = cmd.status()?;
    if !status.success() {
        return Err(format!("Script '{}' failed with exit code: {:?}", script.name, status.code()).into());
    }

    Ok(())
}

fn build_python_command(
    config: &ExecutorConfig,
    script: &ScriptConfig,
    args: &[String],
    working_dir: &PathBuf,
) -> Result<Command, Box<dyn std::error::Error>> {
    let python_exe = if let Some(env_path) = &config.python_env {
        let venv_path = PathBuf::from(env_path);
        
        // Windows: Scripts\python.exe
        let win_python = venv_path.join("Scripts").join("python.exe");
        if win_python.exists() {
            win_python
        } else {
            // Linux/Mac: bin/python
            let unix_python = venv_path.join("bin").join("python");
            if unix_python.exists() {
                unix_python
            } else {
                PathBuf::from(&config.python_path)
            }
        }
    } else {
        PathBuf::from(&config.python_path)
    };

    let mut cmd = Command::new(&python_exe);
    cmd.arg(&script.script_path);
    
    for arg in &script.args {
        cmd.arg(arg);
    }
    
    for arg in args {
        cmd.arg(arg);
    }

    Ok(cmd)
}

fn build_powershell_command(
    config: &ExecutorConfig,
    script: &ScriptConfig,
    args: &[String],
    _working_dir: &PathBuf,
) -> Result<Command, Box<dyn std::error::Error>> {
    let mut cmd = Command::new(&config.powershell_path);
    cmd.arg("-File");
    cmd.arg(&script.script_path);
    
    for arg in &script.args {
        cmd.arg(arg);
    }
    
    for arg in args {
        cmd.arg(arg);
    }

    Ok(cmd)
}

